### PR TITLE
Introducing OVN as oVirt's network provider

### PR DESCRIPTION
**Pull request by AlonaKaplan**
Discovering OVN as the network provider of oVirt and supporting showing
ports and CRUD actions for networks and subnets.
Manageiq will communicate directly with ovirt-provider-ovn.
Since ovirt-provider-ovn has neutron like api as much as possible
openstack's network provider code was reused.

Feature page -
https://github.com/oVirt/ovirt-site/pull/1068

### CODE REVIEW

**Comment by AlonaKaplan** [comment]
@dankenigsberg

**Comment by AlonaKaplan** [comment]
please review @tzumainn @Ladas  @masayag

File: app/models/manageiq/providers/redhat/infra_manager.rb [hunk 1]
```diff
@@ -15,9 +15,42 @@ class ManageIQ::Providers::Redhat::InfraManager < ManageIQ::Providers::InfraMana
...
+
   supports :provisioning
   supports :refresh_new_target
 
+  before_update :ensure_managers
+  before_update :ensure_managers_zone_and_provider_region
+
+  def ensure_network_manager
+    providers = ovirt_services.collect_external_network_providers
+    providers = providers.sort_by(&:name)
```
Comment (Ladas) [inline app/models/manageiq/providers/redhat/infra_manager.rb]: seems like providers are nil in the specs?

File: app/models/manageiq/providers/redhat/infra_manager.rb [hunk 2]
```diff
@@ -15,9 +15,46 @@ class ManageIQ::Providers::Redhat::InfraManager < ManageIQ::Providers::InfraMana
...
 
+  has_many :cloud_tenants, :foreign_key => :ems_id, :dependent => :destroy
+
+  include HasNetworkManagerMixin
+
   supports :provisioning
   supports :refresh_new_target
 
+  before_update :ensure_managers
+  before_update :ensure_managers_zone_and_provider_region
```
Comment (Ladas) [inline app/models/manageiq/providers/redhat/infra_manager.rb]: this line is already called inside ensure managers, we can delete it

File: app/models/manageiq/providers/redhat/infra_manager.rb [hunk 3]
```diff
@@ -15,9 +15,48 @@ class ManageIQ::Providers::Redhat::InfraManager < ManageIQ::Providers::InfraMana
...
+
+  def ensure_network_manager
+    providers = ovirt_services.collect_external_network_providers
+
+    unless providers.blank?
+      providers = providers.sort_by(&:name)
```
Comment (masayag) [inline app/models/manageiq/providers/redhat/infra_manager.rb]: Can we consider also other properties of the provider such as: status or last access time to select the better match ?

Comment (AlonaKaplan) [reply app/models/manageiq/providers/redhat/infra_manager.rb]: Currently, oVirt api doesn't provide us any of these proposed properties.

File: app/models/manageiq/providers/redhat/network_manager.rb [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  include SupportsFeatureMixin
+
+  supports :create
+
+  has_many :public_networks,  :foreign_key => :ems_id, :dependent => :destroy,
+           :class_name => "ManageIQ::Providers::Openstack::NetworkManager::CloudNetwork::Public"
+  has_many :private_networks, :foreign_key => :ems_id, :dependent => :destroy,
+           :class_name => "ManageIQ::Providers::Openstack::NetworkManager::CloudNetwork::Private"
+
+  # Auth and endpoints delegations, editing of this type of manager must be disabled
```
Comment (masayag) [inline app/models/manageiq/providers/redhat/network_manager.rb]: pls rephrase this message - it is a bit unclear

File: app/models/manageiq/providers/redhat/network_manager/refresher.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+        target_name = target.try(:name) || target.try(:event_type)
+
+        _log.info("Filtering inventory for #{target.class} [#{target_name}] id: [#{target.id}]...")
+
+        if ::Settings.ems.ems_ovirt.ems_network.try(:refresh).try(:inventory_object_refresh)
+          inventory = ManageIQ::Providers::Redhat::Builder.build_inventory(ems, target)
+        end
+
+        _log.info("Filtering inventory...Complete")
+        [target, inventory]
```
Comment (masayag) [inline app/models/manageiq/providers/redhat/network_manager/refresher.rb]: What would be the result if line 9 returns false and not value is set to `inventory` ?

File: app/models/manageiq/providers/redhat/infra_manager.rb [hunk 6]
```diff
@@ -15,9 +15,48 @@ class ManageIQ::Providers::Redhat::InfraManager < ManageIQ::Providers::InfraMana
...
+        build_network_manager(:type => 'ManageIQ::Providers::Redhat::NetworkManager')
+      end
+      uri = URI.parse(auth_url)
+      parse_network_manager_url(uri)
+    elsif network_manager
+      network_manager.orchestrate_destroy
+    end
+  end
+
+  def parse_network_manager_url(uri)
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply app/models/manageiq/providers/redhat/infra_manager.rb]: done

File: app/models/manageiq/providers/redhat/network_manager.rb (hunk 4 above)
Comment (AlonaKaplan) [reply app/models/manageiq/providers/redhat/network_manager.rb]: Removed the comment. It was copied from the openstack network manager and not relevant here. Thanks.

File: app/models/manageiq/providers/redhat/network_manager/refresher.rb (hunk 5 above)
Comment (AlonaKaplan) [reply app/models/manageiq/providers/redhat/network_manager/refresher.rb]: The refresh of the 'network manager' will fail.

File: app/models/manageiq/providers/redhat/infra_manager.rb [hunk 7]
```diff
@@ -15,9 +15,60 @@ class ManageIQ::Providers::Redhat::InfraManager < ManageIQ::Providers::InfraMana
...
 
+  has_many :cloud_tenants, :foreign_key => :ems_id, :dependent => :destroy
+
+  include HasNetworkManagerMixin
+
   supports :provisioning
   supports :refresh_new_target
 
+  before_save :ensure_managers
+  before_save :ensure_managers_zone_and_provider_region
```
Comment (Ladas) [inline app/models/manageiq/providers/redhat/infra_manager.rb]: we can delete this line, it is being called inside ensure_managers

**Comment by AlonaKaplan** [comment]
depends on  https://github.com/ManageIQ/manageiq/pull/15929

**Pull request merged**